### PR TITLE
#21 version checker, #23 Callback for Gandalf

### PR DESCRIPTION
**Pull request by jsibbold**
gotta get used to this style, added the Gandalf callback without really thinking, shouldn't be a huge deal. The one area that is sort of strange is the logic involved in the Alert, and it's mostly related to the blocking mechanic. It's the only part that will show regardless of matching the previously seen. We can discuss that logic if we want to change it or remove the Alert and put it in its own set of checks.

### CODE REVIEW

File: gandalf/src/main/java/io/github/btkelly/gandalf/Callback.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.github.btkelly.gandalf;
+
+import io.github.btkelly.gandalf.models.Alert;
+import io.github.btkelly.gandalf.models.OptionalUpdate;
+import io.github.btkelly.gandalf.models.RequiredUpdate;
+
+public interface Callback {
```
Comment (btkelly) [inline gandalf/src/main/java/io/github/btkelly/gandalf/Callback.java]: Maybe we could give this a little more descriptive name? Thoughts?

File: gandalf/src/main/java/io/github/btkelly/gandalf/checker/GateKeeper.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     * Checks version of installed app against the {@link Bootstrap}.
+     * @param bootstrap - provided version requirements
+     * @return true if an update is available and hasn't been recorded in history
+     */
+    public boolean updateIsOptional(@NonNull final Bootstrap bootstrap) {
+        final OptionalUpdate optionalUpdate = bootstrap.getOptionalUpdate();
+
+        return optionalUpdate != null
+                && versionChecker.showOptionalUpdate(optionalUpdate, appVersionDetails)
+                && !historyChecker.contains(optionalUpdate);
```
Comment (btkelly) [inline gandalf/src/main/java/io/github/btkelly/gandalf/checker/GateKeeper.java]: Maybe swap the order of these checks? That way if the check is in the history we can skip the check if we should.

**Comment by btkelly** [comment]
Overall looks good, just a few minor comments

File: gandalf/src/main/java/io/github/btkelly/gandalf/checker/GateKeeper.java (hunk 2 above)
Comment (jsibbold) [inline gandalf/src/main/java/io/github/btkelly/gandalf/checker/GateKeeper.java]: will do

File: gandalf/src/main/java/io/github/btkelly/gandalf/Callback.java (hunk 1 above)
Comment (jsibbold) [inline gandalf/src/main/java/io/github/btkelly/gandalf/Callback.java]: I'm open to anything as far as a name, I think I just wasn't feeling creative when I made the class. And I don't have any ideas at the very moment. I suppose I'll sleep on it and see if I dream up a better name.

**Pull request merged**